File: 02/rust/src/main.rs
use std::fs;

enum Direction {
    Up,
    Down,
    Forward,
}

struct Command {
    direction: Direction,
    units: u32,
}

#[derive(Clone, Copy, Debug)]
struct Vector {
    distance: u32,
    depth: u32,
}

fn parse_command(line: &str) -> Command {
    let parts: Vec<&str> = line.split(" ").collect();
    assert!(parts.len() == 2);
    let direction = match parts[0] {
        "up" => Direction::Up,
        "down" => Direction::Down,
        "forward" => Direction::Forward,
        _ => panic!("Can't parse {}", parts[0]),
    };
    let units: u32 = parts[1].parse().unwrap();
    Command { direction, units }
}

fn read_puzzle_input(filename: &str) -> Vec<Command> {
    let content = fs::read_to_string(filename).expect("Couldn't read input file");
    content.lines().map(|x| parse_command(x)).collect()
}

fn follow_course(course: &Vec<Command>) -> Vector {
    let mut location = Vector {
        distance: 0,
        depth: 0,
    };
    for command in course {
        match command.direction {
            Direction::Up => location.depth -= command.units,
            Direction::Down => location.depth += command.units,
            Direction::Forward => location.distance += command.units,
        }
    }
    location
}

fn follow_course_correctly(course: &Vec<Command>) -> Vector {
    let mut location = Vector {
        distance: 0,
        depth: 0,
    };
    let mut aim = 0;
    for command in course {
        match command.direction {
            Direction::Up => aim -= command.units,
            Direction::Down => aim += command.units,
            Direction::Forward => {
                location.distance += command.units;
                location.depth += aim * command.units;
            }
        }
    }
    location
}

fn main() {
    let puzzle_input = read_puzzle_input("../input.txt");
    let destination = follow_course(&puzzle_input);
    println!(
        "{:?} => {}",
        destination,
        destination.distance * destination.depth
    );
    let destination = follow_course_correctly(&puzzle_input);
    println!(
        "{:?} => {}",
        destination,
        destination.distance * destination.depth
    );
}

File: 02/main.hs
data Position = Position Int Int deriving (Show)
data Direction = Forward | Up | Down
data Command = Command Direction Int

parseCommand :: String -> Command
parseCommand ('f' : 'o' : 'r' : 'w' : 'a' : 'r' : 'd' : ' ' : rest) = Command Forward (read rest)
parseCommand ('u' : 'p' : ' ' : rest) = Command Up (read rest)
parseCommand ('d' : 'o' : 'w' : 'n' : ' ' : rest) = Command Down (read rest)

followCourse :: Position -> [Command] -> Position
followCourse p [] = p
followCourse (Position dist depth) (c : cs) =
    case c of
        (Command Forward x) -> followCourse (Position (dist + x) depth) cs
        (Command Up x) -> followCourse (Position dist (depth - x)) cs
        (Command Down x) -> followCourse (Position dist (depth + x)) cs

followCourseCorrectly :: Position -> Int -> [Command] -> Position
followCourseCorrectly p a [] = p
followCourseCorrectly p@(Position dist depth) aim (c : cs) =
    case c of
        (Command Forward x) -> followCourseCorrectly (Position (dist + x) (depth + aim * x)) aim cs
        (Command Up x) -> followCourseCorrectly p (aim - x) cs
        (Command Down x) -> followCourseCorrectly p (aim + x) cs

main :: IO ()
main = do
    content <- readFile "input.txt"
    let commands = map parseCommand (lines content)
    print (followCourse (Position 0 0) commands)
    print (followCourseCorrectly (Position 0 0) 0 commands)
